pwnlib/fmtstr.py

```diff
@@ -102,7 +108,497 @@ def send_payload(payload):
+            (carry, strict): find_min_hamming_in_range_step(prev, step, carry, strict )
+            for carry in [False, True]
+            for strict in [False, True]
+        }
+    return prev[(False,False)]
+#
+# what we don't do:
+#  - create new atoms that cannot be created by merging existing atoms
+#  - optimize based on masks
+def merge_atoms_overlapping(atoms, sz, szmax, numbwritten, overflows):
+    if not szmax:
+        szmax = max(SPECIFIER.keys())
+
+    assert 1 <= overflows, "must allow at least one overflow"
+    assert sz <= szmax, "sz must be smaller or equal to szmax"
+
+    maxwritten = numbwritten + (1 << (8 * sz)) * overflows
+    done = [False for _ in atoms]
+
+    def possible_writes(idx, numbwritten):
+        candidate = AtomWrite(atoms[idx].start, 0, 0)
+        best = 0
+        for nextidx in xrange(idx, len(atoms)):
+            if candidate.size > szmax or done[nextidx] or candidate.end != atoms[nextidx].start:
+                break
+
+            candidate = candidate.union(atoms[nextidx])
+            if candidate.size not in SPECIFIER: continue
+
+            approxed = candidate
+            score = candidate.size
+            if approxed.size > sz:
+                score, v, m = find_min_hamming_in_range(approxed.size, numbwritten, maxwritten, approxed.integer)
+                approxed = candidate.replace(integer=v, mask=m)
+
+            if score > best:
+                best = score
+                yield list(xrange(idx, nextidx+1)), approxed
+
+    numbwritten_at = [numbwritten for _ in atoms]
+    out = []
+    for idx in xrange(len(atoms)):
+        if done[idx]: continue
+        numbwritten_here = numbwritten_at[idx]
+        idxrange, candidate = list(possible_writes(idx, numbwritten_here))[-1]
+        numbwritten_here += candidate.compute_padding(numbwritten_here)
+        offset = 0
+        for i in idxrange:
+            shift = atoms[i].size
+            if not ((~candidate[offset:offset+shift].mask) & atoms[i].mask):
+                done[i] = True
+            else:
+                numbwritten_at[i] = max(numbwritten_at[i], numbwritten_here)
+            offset += shift
+        out += [candidate]
+    return out
+
+def sort_atoms(atoms, numbwritten):
+    # find dependencies
+    order = { atom: i for i,atom in enumerate(atoms) }
+
+    def conflicts():
+        prev = None
+        for a in sorted(atoms, key=lambda a: a.start):
+            if not prev:
+                prev = a
+                continue
+            if prev.end > a.start:
+                yield prev, a
+            if a.end > prev.end:
+                prev = a
+
+    depgraph = { a: set() for a in atoms }
+    rdepgraph = { a: set() for a in atoms }
+    for a,b in conflicts():
+        if order[a] < order[b]:
+            depgraph[b].add(a)
+            rdepgraph[a].add(b)
+        else:
+            depgraph[a].add(b)
+            rdepgraph[b].add(a)
+
+    szmask = { sz: (1 << (sz * 8)) - 1 for sz in SPECIFIER }
```

Comment:
`szmask`, `queues`, and `pointers` all look like constants.  Can you promote them to module-level and make them allcaps?

Reply:
only szmask is a constant here. I will add a few more comments and docs to this function